yb-voyager/cmd/exportData.go

```diff
@@ -788,14 +855,258 @@ func getInitialTableList() (map[string]string, []sqlname.NameTuple) {
+				return nil, nil, fmt.Errorf("failed to add the leaf partitions in table list: %w", err)
+			}
+		}
+	}
+	return firstRunTableWithLeafsAndRoots, partitionsToRootTableMap, nil
+}
+
+func getInitialTableList() (map[string]string, []sqlname.NameTuple, error) {
+	// store table list after filtering unsupported or unnecessary tables
+	msr, err := metaDB.GetMigrationStatusRecord()
+	if err != nil {
+		utils.ErrExit("error fetching migration status record: %v", err)
+	}
+	storedTableListNotAvailable := func() bool {
+		//if any of the list in DB is empty
+		return len(msr.TableListExportedFromSource) == 0 || len(msr.SourceExportedTableListWithLeafPartitions) == 0
 	}
 
-	return partitionsToRootTableMap, finalTableList
+	if bool(startClean) || storedTableListNotAvailable() {
+		// fresh start case or the first run where we don't have a table list stored in msr
+		return fetchTablesNamesFromSourceAndFilterTableList()
+
+	}
+
+	//sunsequent run case where we will use a table-list stored in msr
+	var firstRunTableWithLeafsAndRoots []sqlname.NameTuple
+	var partitionsToRootTableMap map[string]string
+
+	firstRunTableWithLeafsAndRoots, partitionsToRootTableMap, err = retrieveFirstRunListAndPartitionsRootMap(msr)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting the first run table list and partition to root mapping: %v", err)
+	}
+
+	//guardrails around the table-list in case of re-run
+
+	registeredList, err := getRegisteredNameRegList()
+	if err != nil {
+		return nil, nil, fmt.Errorf("error getting name registry list: %v", err)
+	}
+
+	rootTables := make([]sqlname.NameTuple, 0)
+	for _, v := range partitionsToRootTableMap {
+		tuple, err := namereg.NameReg.LookupTableName(v)
+		if err != nil {
+			utils.ErrExit("look up failed for the table name: %s: %v", v, err)
+		}
+		rootTables = append(rootTables, tuple)
+	}
+	rootTables = lo.UniqBy(rootTables, func(t sqlname.NameTuple) string {
+		return t.ForKey()
+	})
+
+	// Finding all the partitions of all root tables part of migration, and report if there any new partitions added
+	rootToNewLeafTablesMap, err := detectNewLeafPartitionsOnPartitionedTables(rootTables, registeredList)
+	if err != nil {
+		return nil, nil, fmt.Errorf("detecting new leaf tables on the partitioned tables: %v", err)
+	}
+
+	if source.TableList == "" && source.ExcludeTableList == "" {
+		//which mean no table-lists are passed in subsequent runs
+		// so no need to do the guardrails check and all
+		return partitionsToRootTableMap, firstRunTableWithLeafsAndRoots, nil
+	}
+
+	firstRunTableWithLeafParititons, currentRunTableListWithLeafPartitions, err := applyTableListFlagsOnCurrentAndRemoveRootsFromBothLists(registeredList, source.TableList, source.ExcludeTableList, rootToNewLeafTablesMap, rootTables, firstRunTableWithLeafsAndRoots)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error applying table list flags for current list and remove roots: %v", err)
+	}
+
+	if len(lo.Keys(rootToNewLeafTablesMap)) > 0 {
```

Comment:
shouldn't this check be done before the `source.TableList == "" && source.ExcludeTableList == ""` if block?

Reply:
moved it in the `detectNewLeafPartitionsOnPartitionedTable` function back.